### PR TITLE
Refactoring #2

### PR DESCRIPTION
**Pull request by ivan-kleshnin**
Implementation of https://github.com/jurassix/react-validation-mixin/pull/8
See also https://github.com/jurassix/react-validation-mixin/pull/10

### CODE REVIEW

File: ValidationMixin.js [hunk 1]
```diff
@@ -44,18 +43,40 @@ var ValidationMixin = {
...
-   * @param {?String} State key to validate
+   * Usage: <input onBlur={this.handleUnfocusFor('password')} .../>
+   *
+   * @param {?String} key to validate
    * default is false.
    * @return {function} validation event handler
    */
-  handleValidation: function(key) {
-    return function(event) {
+  handleUnfocusFor: function(key) {
```
Comment (jurassix) [inline ValidationMixin.js]: `handleUnfocusFor` does not account for the fact this event handler could be attached to an `onChange` event or really any event. I think we need to stick with the generic `handleValidation` or something like it to emphasize this fact.

File: ValidationMixin.js [hunk 2]
```diff
@@ -44,18 +43,40 @@ var ValidationMixin = {
...
       this.validate(key);
     }.bind(this);
   },
 
   /**
+   * Convenience method to validate a whole form on submit
+   *
+   * Usage: <form onSubmit={this.handleSubmit}>...</form>
+   */
+  handleSubmit: function(event) {
```
Comment (jurassix) [inline ValidationMixin.js]: I'm not sure about this from a developers perspective, we cannot generically handle the submit without providing a hook or callback to execute if validation is successful. I have an idea about how to handle this but am still working it out. I'll push it soon at let you review.

File: ValidationMixin.js [hunk 3]
```diff
@@ -44,18 +43,40 @@ var ValidationMixin = {
...
+    event.preventDefault();
+    this.validate();
+  },
+
+  /**
+   * Convenience method to reset a form to initial state
+   *
+   * Usage: <button onClick={this.handleReset}>Reset</button>
+   */
+  handleReset: function(event) {
```
Comment (jurassix) [inline ValidationMixin.js]: `handleReset` is a good suggestion for this library. I think I'll add `clearValidations` to the API that will simply clear out `this.state.errors`

File: ValidationMixin.js [hunk 4]
```diff
@@ -72,11 +93,7 @@ var ValidationMixin = {
    * @return {Boolean}.
    */
   isValid: function(key) {
```
Comment (jurassix) [inline ValidationMixin.js]: I think we definitely need `isValid` to be implemented this way; but I first need to expose an API to give the user a way to get the same behavior. I'll put something out this week to discuss.

**Comment by jurassix** [comment]
Thanks for your feedback Ivan. Always appreciated. I'll try to get some commits in this week to clean up the API and account for some of the discrepancies we have discussed.

**Comment by ivan-kleshnin** [comment]
I've added you as collaborator to my fork, so you may push right there and the changes will be reflected in this PR.

**Comment by jurassix** [comment]
I pushed my 4.0.0 release to ivan-kleshnin-master. I have solidified the API. Here are the release notes:

4.0.0 (Major release with breaking API changes):

  * `validatorData` is now `getValidatorData` to be more idiomatic of React
  * `validate` now takes an error-first callback, which allows for a simpler API when validating on Submit. `validate` also takes an optional key as a first param that can be used to validate a single form field.
  * `isValid` is now a simple wrapper around `this.state.errors` **see example for handling form submission in 4.x.x**
  * `clearValidations` has been added to reset all current errors.
  * `handleValidations` now takes a second parameter error-first callback that can be used for custom handling of errors.


I'm going to release this version major tonight. I'll also update the `gh-pages` to match these changes.

Next I'll get the compound-keys #13 PR updated to match master and get it merged into 4.x.x

Let me know if there are any major objections to the next release. I've handled the `isValid` and the `onSubmit` cases to be clear. I'm adopting a node error-first callback style for `validate` calls.

**Comment by ivan-kleshnin** [comment]
Everything looks solid. I'm a bit puzzled by "validations" vs "errors" distinction. Is there any?

**Comment by jurassix** [comment]
By `validations` vs `errors` are your referring to this code in the demo?

```javascript
    onValidate = function(error, validationErrors) {
      if (error) {
        this.setState({
          feedback: 'Form is invalid do not submit'
        });
      } else {
        this.setState({
          feedback: 'Form is valid send to action creator'
        });
      }
    }.bind(this);
    this.validate(onValidate);
```

Basically the second parameter, `validationErrors` is just the errors object returned by `ValidationFactory.validate`; it's equivalent to `this.state.errors` but is left in as a convenience. The real useful information to the user is `errors` parameter, which allows node style checking if there were errors during validation.

I've also beefed up the testing harness with the **4.0.3** release. I'll add a couple more UI tests to ensure any changes we make to the Factory doesn't break the users expectations. I'll also be working on merging in #11 after I have some time to digest and review.

**Comment by ivan-kleshnin** [comment]
I mean pure naming aspect. Somewhere it's `errors` like in `this.state.errors`, somewhere it's `validations` like in `resetValidations` / `handleValidations`, somewhere its 'validationErrors' like in example above. Why not just `errors` so `resetErrors` / `handleErrors` / `errors` arg / `this.state.errors`.

**Pull request merged**